include/xtensor/xsemantic.hpp

```diff
@@ -18,8 +18,211 @@
+                : m_first(reinterpret_cast<uintptr_t>(last < first ? last : first)),
+                m_last(reinterpret_cast<uintptr_t>(last < first ? first : last)) {}
+
+            template <class T>
+            bool overlaps(T* first, T* last) const
+            {
+                if (first <= last)
+                {
+                    return reinterpret_cast<uintptr_t>(first) <= m_last && reinterpret_cast<uintptr_t>(last) >= m_first;
+                }
+                else
+                {
+                    return reinterpret_cast<uintptr_t>(last) <= m_last && reinterpret_cast<uintptr_t>(first) >= m_first;
+                }
+            }
+        };
+
+        template <class E, class Enable = void>
+        struct overlapping_memory_checker_traits
+        {
+            static bool check_overlap(const E&, const memory_range&)
+            {
+                return true;
+            }
+        };
+
+        template <class E>
+        struct overlapping_memory_checker_traits<E, std::enable_if_t<has_memory_address<E>::value>>
+        {
+            static bool check_overlap(const E& expr, const memory_range& dst_range)
+            {
+                if (expr.size() == 0)
+                {
+                    return false;
+                }
+                else
+                {
+                    return dst_range.overlaps(std::addressof(*expr.begin()), std::addressof(*expr.rbegin()));
+                }
+            }
+        };
+
+        template <class E>
+        struct overlapping_memory_checker_traits<E, std::enable_if_t<!has_memory_address<E>::value &&
+                                                                    is_crtp_base_of<xview_semantic, E>::value>>
+        {
+            static bool check_overlap(const E& expr, const memory_range& dst_range)
+            {
+                if (expr.size() == 0)
+                {
+                    return false;
+                }
+                else
+                {
+                    using ChildE = std::decay_t<decltype(expr.expression())>;
+                    return overlapping_memory_checker_traits<ChildE>::check_overlap(expr.expression(), dst_range);
+                }
+            }
+        };
+
+        template <class E>
+        struct overlapping_memory_checker_traits<E, std::enable_if_t<!has_memory_address<E>::value &&
+                                                                    is_specialization_of<xbroadcast, E>::value>>
+        {
+            static bool check_overlap(const E& expr, const memory_range& dst_range)
+            {
+                if (expr.size() == 0)
+                {
+                    return false;
+                }
+                else
+                {
+                    using ChildE = std::decay_t<decltype(expr.expression())>;
+                    return overlapping_memory_checker_traits<ChildE>::check_overlap(expr.expression(), dst_range);
+                }
+            }
+        };
```

Comment:
This specialization should be defined in `xbroadcast.hpp` to avoid "circular" dependencies.